crates/proto/build.rs

```diff
@@ -3,53 +3,116 @@ use std::{env, fs, path::PathBuf};
+    let compiled_path: PathBuf = env::var("OUT_DIR").into_diagnostic()?.into();
+    let compiled_path = compiled_path.join("file_descriptor_set.bin");
+    fs::write(&compiled_path, compiled_data).into_diagnostic()?;
 
     let mut prost_config = prost_build::Config::new();
     prost_config.skip_debug(["AccountId", "Digest"]);
 
     // Generate the stub of the user facing server from its proto file
     tonic_build::configure()
-        .file_descriptor_set_path(&file_descriptor_path)
+        .file_descriptor_set_path(compiled_path)
         .skip_protoc_run()
-        .out_dir("src/generated")
+        .out_dir(&target_dir)
         .compile_with_config(prost_config, protos, includes)
         .into_diagnostic()?;
 
+    // Generate the module file and other miscellaneous files.
+    let generated_files = file_stems(&target_dir).into_diagnostic()?;
+    generate_mod_file(target_dir, &generated_files).into_diagnostic()?;
+    generate_proto_file_getter().into_diagnostic()?;
+
+    Ok(())
+}
+
+/// Sorted list of the file stems (file names without the extension).
+fn file_stems(target_dir: &PathBuf) -> std::io::Result<Vec<String>> {
+    let mut filestems = Vec::new();
+
+    for entry in fs::read_dir(target_dir)? {
+        let entry = entry?;
+        let path = entry.path();
+        if path.is_file() {
+            let filename = path
+                .file_stem()
+                .expect("Filename")
+                .to_str()
+                .expect("Filename should be utf8")
+                .to_owned();
+
+            filestems.push(filename);
+        }
+    }
+
+    filestems.sort();
+
+    Ok(filestems)
+}
+
+/// Generate a mod.rs which includes all the provided submodules.
+fn generate_mod_file(target_dir: PathBuf, submodules: &[String]) -> std::io::Result<()> {
+    let content = submodules.iter().map(|f| format!("pub mod {f};\n"));
+    let content = std::iter::once("// Generated by build.rs\n".to_owned())
+        .chain(content)
+        .collect::<String>();
+
+    fs::write(target_dir.join("mod.rs"), content)?;
+
+    Ok(())
+}
+
+fn generate_proto_file_getter() -> std::io::Result<()> {
+    // Unfortunately doing relative paths is convoluted, so this function harcodes
+    // the paths instead.
+
+    use std::io::Write;
+
+    let mut target = fs::File::create("src/proto_files.rs")?;
+
+    writeln!(target, "// File generated by buid.rs")?;
+    writeln!(target)?;
+    writeln!(target, "/// List of protobuf files and their contents")?;
+    writeln!(target, "pub const PROTO_FILES: &[(&str, &str)] = &[")?;
+
+    for entry in std::fs::read_dir("proto")? {
+        let entry = entry?;
+        let path = entry.path();
+        if path.is_file() {
+            let filename = path
+                .file_name()
+                .and_then(|f| f.to_str())
+                .expect("Filename should exist and  be utf8");
+            writeln!(target, r#"    ("{filename}.rs", include_str!("../proto/{filename}")),"#)?;
```

Comment:
Is the `.rs` suffix in `{filename}.rs` correct? The map would then have keys such as `requests.proto.rs`

Reply:
Thanks; I think I've successfully confused myself with strings and extensions.